features/26-mfa/api.md

```diff
@@ -0,0 +1,975 @@
+  # Default is false.
+  enabled: false
+  # Valid values are 'off', 'optional' or 'required'.
+  # Default is 'optional'.
+  #
+  # 'off' means MFA is never required during authentication.
+  #
+  # 'optional' means MFA is required during authentication when the user has at least 1 activated authenticator. This is like how other major online services such as Google offer MFA.
+  #
+  # 'required' means MFA is always required during authentication.
+  enforcement: 'optional'
+  # The total maximum number of authenticator the user can register.
+  # Valid values is [0, 15].
+  # Default is 1.
+  maximum: 1
+  totp:
+    # The maximum number of TOTP authenticator the user can register.
+    # Valid values is [0, 5].
+    # Default is 0.
+    maximum: 0
+  oob:
+    sms:
+      # The maximum number of OOB-SMS authenticator the user can register.
+      # Valid values is [0, 5].
+      # Default is 0.
+      maximum: 0
+    email:
+      # The maximum number of OOB-Email authenticator the user can register.
+      # Valid values is [0, 5].
+      # Default is 0.
+      maximum: 0
+  bearer_token:
+    # How many days the bearer token is valid. Default is 30.
+    expire_in_days: 30
+  recovery_code:
+    # How many recovery code should be generated. Default is 16.
+    count: 16
+    # Enable listing the existing recovery code.
+    # Some services like GitHub allow listing the existing recovery code.
+    # By default listing is disabled, the user must regenerate a new set of
+    # recovery code.
+    list_enabled: false
+```
+
+# Configuration Recipes
+
+## At most 1 TOTP or 1 OOB-SMS or 1 OOB-Email
+
+```yaml
+mfa:
+  enabled: true
+  enforcement: "optional"
+  totp:
+    maximum: 1
+  oob:
+    sms:
+      maximum: 1
+    email:
+      maximum: 1
+```
+
+## At most 1 TOTP and 1 OOB-SMS and 1 OOB-Email
+
+```yaml
+mfa:
+  enabled: true
+  enforcement: "optional"
+  maximum: 3
+  totp:
+    maximum: 1
+  oob:
+    sms:
+      maximum: 1
+    email:
+      maximum: 1
+```
+
+## At most 1 TOTP
+
+```yaml
+mfa:
+  enabled: true
+  enforcement: "optional"
+  totp:
+    maximum: 1
+```
+
+## At most 1 OOB-SMS
+
+```yaml
+mfa:
+  enabled: true
+  enforcement: "optional"
+  oob:
+    sms:
+      maximum: 1
+```
+
+# Endpoints
+
+## POST /mfa/totp/new
+
+### Request body
+
+```JSON
+{
+  "display_name": "my TOTP",
+  "account_name": "user@example.com",
+  "issuer": "MyApp"
+}
+```
+
+- `display_name`: The display name of the new TOTP authenticator.
+- `account_name`: If this is provided, it is used to generate `otpauth_uri`.
+- `issuer`: If this is provided, it is used to generate `otpauth_uri`.
+- `authn_session_token`: Required only when there is no access token.
+
+### Response body
+
+```JSON
+{
+  "authenticator_id": "xxx",
+  "authenticator_type": "totp",
+  "secret": "secret",
+  "otpauth_uri": "otpauth://totp/MyApp:user@example.com?secret=secret&issuer=MyApp"
```

Comment:
Following up from the comment of avoiding secret in GET parameter as part of URL, do we really need to send the secret as string to client side at all? Maybe make secret "optional" and only return in response if the client want it? In normal flow I think the client side only need the QR Code (which itself have the secrets already)

I might have missed something! Just brainstorming...